### PR TITLE
[WIP] Support for concat option in C4 Dataset

### PR DESCRIPTION
**Pull request by karan6181**
## Description of changes:
- Support for concat option in C4 Dataset. Reference from https://github.com/mosaicml/benchmarks/blob/main/llm/src/data_c4.py.
- Added a readme on how to use `c4.py` script to convert the raw dataset into a `.mds` format. Reference from [mosaicml/benchmarks](https://github.com/mosaicml/benchmarks/tree/main/llm#dataset-preparation) repo.

## Issue #, if available:
https://github.com/mosaicml/streaming/issues/76

## Merge Checklist:
_Put an `x` without space in the boxes that apply. If you are unsure about any checklist, please don't hesitate to ask. We are here to help! This is simply a reminder of what we are going to look for before merging your pull request._

### General
- [x] I have read the [contributor guidelines](https://github.com/mosaicml/streaming/blob/main/CONTRIBUTING.md)
- [ ] This is a documentation change or typo fix. If so, skip the rest of this checklist.
- [x] I certify that the changes I am introducing will be backward compatible, and I have discussed concerns about this, if any, with the MosaicML team.
- [x] I have updated any necessary documentation, including [README](https://github.com/mosaicml/streaming/blob/main/README.md) and [API docs](https://github.com/mosaicml/streaming/tree/main/docs) (if appropriate).

### Tests
- [x] I ran `pre-commit` on my change. (check out the `pre-commit` section of [prerequisites](https://github.com/mosaicml/streaming/blob/main/CONTRIBUTING.md#prerequisites))
- [ ] I have added tests that prove my fix is effective or that my feature works (if appropriate).
- [ ] I ran the tests locally to make sure it pass. (check out [testing](https://github.com/mosaicml/streaming/blob/main/CONTRIBUTING.md#running-tests))
- [ ] I have added unit and/or integration tests as appropriate to ensure backward compatibility of the changes.

### CODE REVIEW

File: streaming/text/c4.py [hunk 1]
```diff
@@ -105,3 +108,47 @@ def __getitem__(self, idx: int) -> Any:
...
+        """
+        if self.group_method == 'truncate':
+            iterator = super().__iter__()
+            yield from iterator
+        else:  # group_method = 'concat'
+            buffer = {}
+            while True:
+                iterator = super().__iter__()
+                for sample in iterator:
+
```
Comment (knighton) [inline streaming/text/c4.py:132]: spurious newline

File: streaming/text/c4.py [hunk 2]
```diff
@@ -105,3 +108,47 @@ def __getitem__(self, idx: int) -> Any:
...
+        parent iterator class.
+
+        For `group_method = truncate`, simply return the token sample.
+        For `group_method = concat`, keep fetching token samples until it fills up the max_seq_len.
+
+        Yields:
+            Iterator[Any]: Sample iterator
+        """
+        if self.group_method == 'truncate':
+            iterator = super().__iter__()
```
Comment (knighton) [inline streaming/text/c4.py:125]: yield from super().__iter__()?

File: streaming/text/convert/README.md [hunk 3]
```diff
@@ -0,0 +1,17 @@
+# Dataset preparation
+
+To use Streaming Dataset, We first convert the dataset from its native format to MosaicML's streaming dataset format (a collection of binary `.mds` files). Once in `.mds` format, we can store the dataset in a central location (filesystem, S3, GCS, etc.) and stream the data to any compute cluster, with any number of devices, and any number of CPU workers, and it all ~ just works ~ .
```
Comment (knighton) [inline streaming/text/convert/README.md:3]: should it be ~just works~?

File: streaming/text/c4.py (hunk 1 above)
Comment (karan6181) [reply streaming/text/c4.py]: Fixed in https://github.com/mosaicml/streaming/pull/77/commits/0a63fc68fb371e4b37fffe4ba043990b3816a9c4.

File: streaming/text/convert/README.md (hunk 3 above)
Comment (karan6181) [reply streaming/text/convert/README.md]: Fixed in https://github.com/mosaicml/streaming/pull/77/commits/0a63fc68fb371e4b37fffe4ba043990b3816a9c4.

File: streaming/text/c4.py (hunk 2 above)
Comment (karan6181) [reply streaming/text/c4.py]: Fixed in https://github.com/mosaicml/streaming/pull/77/commits/0a63fc68fb371e4b37fffe4ba043990b3816a9c4.

**Pull request merged**